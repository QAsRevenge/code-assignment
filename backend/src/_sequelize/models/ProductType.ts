import {
  AutoIncrement,
  Column,
  DataType,
  Model,
  NotEmpty,
  PrimaryKey,
  Table,
} from 'sequelize-typescript';
import {SCHEMA_NAME} from '../constants';

@Table({schema: SCHEMA_NAME.DATA})
class ProductType extends Model {
  @PrimaryKey
  @AutoIncrement
  @Column(DataType.INTEGER)
  declare id: number;

  @NotEmpty
  @Column({type: DataType.STRING, unique: true})
  declare productType: string;
}

export {ProductType};
