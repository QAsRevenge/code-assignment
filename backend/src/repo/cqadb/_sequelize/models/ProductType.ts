import {Column, DataType, Model, NotEmpty, Table} from 'sequelize-typescript';
import {SCHEMA_NAME} from '../constants';

@Table({schema: SCHEMA_NAME.DATA})
class ProductType extends Model {
  @NotEmpty
  @Column({type: DataType.STRING, unique: true})
  declare productType: string;
}

export {ProductType};
