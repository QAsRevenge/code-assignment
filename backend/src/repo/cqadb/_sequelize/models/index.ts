export {ProductType} from './ProductType';
export {Product} from './Product';
export {Colour} from './Colour';
export {ProductColour} from './ProductColour';
