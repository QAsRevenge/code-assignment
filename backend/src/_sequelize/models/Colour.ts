import {
  Table,
  Column,
  Model,
  DataType,
  BelongsToMany,
  AutoIncrement,
  PrimaryKey,
} from 'sequelize-typescript';
import {Product} from './Product';
import {ProductColour} from './ProductColour';
import {SCHEMA_NAME} from '../constants';

@Table({schema: SCHEMA_NAME.DATA})
class Colour extends Model {
  @PrimaryKey
  @AutoIncrement
  @Column(DataType.INTEGER)
  declare id: number;

  @Column({type: DataType.STRING, allowNull: false, unique: true})
  declare name: string;

  @BelongsToMany(() => Product, () => ProductColour)
  declare products: Product[];
}

export {Colour};
