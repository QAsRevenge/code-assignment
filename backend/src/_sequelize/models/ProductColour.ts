import {
  Table,
  Column,
  Model,
  ForeignKey,
  AutoIncrement,
  DataType,
  PrimaryKey,
} from 'sequelize-typescript';
import {Product} from './Product';
import {Colour} from './Colour';
import {SCHEMA_NAME} from '../constants';

@Table({schema: SCHEMA_NAME.DATA})
class ProductColour extends Model {
  @PrimaryKey
  @AutoIncrement
  @Column(DataType.INTEGER)
  declare id: number;

  @ForeignKey(() => Product)
  declare productId: number;

  @ForeignKey(() => Colour)
  declare colourId: number;
}

export {ProductColour};
