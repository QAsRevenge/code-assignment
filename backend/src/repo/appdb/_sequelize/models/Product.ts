import {
  Table,
  Column,
  Model,
  DataType,
  ForeignKey,
  BelongsTo,
  BelongsToMany,
} from 'sequelize-typescript';
import {ProductType} from './ProductType';
import {Colour} from './Colour';
import {ProductColour} from './ProductColour';
import {SCHEMA_NAME} from '../constants';

@Table({schema: SCHEMA_NAME.DATA})
class Product extends Model {
  @Column({type: DataType.STRING, allowNull: false})
  declare name: string;

  @ForeignKey(() => ProductType)
  @Column({type: DataType.INTEGER, allowNull: false})
  declare productTypeId: number;

  @BelongsTo(() => ProductType)
  declare productType: ProductType;

  @BelongsToMany(() => Colour, () => ProductColour)
  declare colours: Colour[];
}

export {Product};
