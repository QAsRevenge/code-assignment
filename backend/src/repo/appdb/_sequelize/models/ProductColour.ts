import {
  Table,
  Column,
  Model,
  ForeignKey,
  AutoIncrement,
  DataType,
  PrimaryKey,
} from 'sequelize-typescript';
import {Product} from './Product';
import {Colour} from './Colour';
import {SCHEMA_NAME} from '../constants';

@Table({schema: SCHEMA_NAME.DATA})
class ProductColour extends Model {
  @PrimaryKey
  @AutoIncrement
  @Column(DataType.INTEGER)
  declare id: number;

  @ForeignKey(() => Product)
  @Column
  productId!: number;

  @ForeignKey(() => Colour)
  @Column
  colourId!: number;
}

export {ProductColour};
