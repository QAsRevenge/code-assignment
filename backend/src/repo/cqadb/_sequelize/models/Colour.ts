import {
  Table,
  Column,
  Model,
  DataType,
  BelongsToMany,
} from 'sequelize-typescript';
import {Product} from './Product';
import {ProductColour} from './ProductColour';
import {SCHEMA_NAME} from '../constants';

@Table({schema: SCHEMA_NAME.DATA})
class Colour extends Model {
  @Column({type: DataType.STRING, allowNull: false, unique: true})
  declare name: string;

  @BelongsToMany(() => Product, () => ProductColour)
  declare products: Product[];
}

export {Colour};
