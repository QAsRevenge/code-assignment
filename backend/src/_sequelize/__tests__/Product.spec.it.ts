import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from 'vitest';
import {Colour, Product, ProductColour, ProductType} from '../models';
import {SqlTestDb, SqlDbTestUtils} from 'test-utils';

describe('Product Model', () => {
  let testDb: SqlTestDb;
  const timeout = 60 * 1000;

  beforeAll(async () => {
    vi.setSystemTime(new Date('2025-01-01'));
    testDb = await SqlDbTestUtils.createDb();

    testDb.sequelize.addModels([Product, ProductColour, ProductType, Colour]);
  }, timeout);

  // This is very hacky and should be done in test-utils but for demonstration purposes..
  beforeEach(async () => {
    await testDb.sequelize.query(
      'TRUNCATE TABLE "data"."product_colours", "data"."products", "data"."colours", "data"."product_types" RESTART IDENTITY CASCADE;',
    );
  }, timeout);

  afterAll(async () => {
    if (testDb) {
      await testDb.close();
    }
  }, timeout);

  it(
    'can create a product',
    async () => {
      // Arrange
      const productType = await ProductType.create({
        productType: 'Test Type',
      });

      const product = await Product.create({
        name: 'Test Product',
        productTypeId: productType.id,
      });

      // Act & Assert
      expect(await Product.findByPk(product.id)).toMatchSnapshot();
    },
    timeout,
  );

  it(
    'cannot create Product with name as null',
    async () => {
      // Act & Assert
      const productType = await ProductType.create({productType: 'Test Type'});
      await expect(
        Product.create({name: null, productTypeId: productType.id}),
      ).rejects.toThrowError('notNull Violation: Product.name cannot be null');
    },
    timeout,
  );

  it(
    'cannot create Product with productTypeId as null',
    async () => {
      // Act & Assert
      await expect(
        Product.create({name: 'Test Product', productTypeId: null}),
      ).rejects.toThrowError(
        /notNull Violation: Product.productTypeId cannot be null/,
      );
    },
    timeout,
  );
});
