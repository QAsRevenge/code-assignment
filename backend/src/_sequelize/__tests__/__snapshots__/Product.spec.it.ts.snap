// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Model > can create a product 1`] = `
{
  "createdAt": 2025-01-01T00:00:00.000Z,
  "id": 1,
  "name": "Test Product",
  "productTypeId": 1,
  "updatedAt": 2025-01-01T00:00:00.000Z,
}
`;
